pkg/resource/deploy/source_eval.go

```diff
@@ -282,15 +311,21 @@ func (d *defaultProviders) newRegisterDefaultProviderEvent(
 //
 // Note that this function must not be called from two goroutines concurrently; it is the responsibility of d.serve()
 // to ensure this.
-func (d *defaultProviders) handleRequest(pkg tokens.Package) (providers.Reference, error) {
-	logging.V(5).Infof("handling default provider request for package %s", pkg)
-
-	ref, ok := d.providers[pkg]
+func (d *defaultProviders) handleRequest(req providers.ProviderRequest) (providers.Reference, error) {
+	logging.V(5).Infof("handling default provider request for package %s", req)
+
+	// Have we loaded this provider before? Use the existing reference, if so.
+	//
+	// Note that we are using the request's String as the key for the provider map. Go auto-derives hash and equality
+	// functions for aggregates, but the one auto-derived for ProviderRequest does not have the semantics we want. The
```

Comment:
In particular - since `ProviderRequest` contains a pointer, Go's hash function is the pointer's has function and not the hash function of the type behind the pointer.